spyder/plugins/editor/widgets/tests/test_bookmarks.py

```diff
@@ -0,0 +1,275 @@
+
+
+@pytest.fixture(scope="module")
+def test_file(tmpdir_factory):
+    """Create and save some python code and text in temporary file."""
+    tmpdir = tmpdir_factory.mktemp("files")
+
+    filename = osp.join(tmpdir.strpath, 'foo1.py')
+    with open(filename, 'w') as f:
+        f.write("# -*- coding: utf-8 -*-\n"
+                "def foo:\n"
+                "    print(Hello World!)\n")
+
+    return filename
+
+
+# --- Tests
+# -----------------------------------------------------------------------------
+def test_add_bookmark(code_editor_bot, mocker):
+    """Test CodeEditor.add_bookmark. Adds bookmark data to Textblock."""
+    editor, _ = code_editor_bot
+    ab = editor.add_bookmark
+
+    mocker.patch.object(codeeditor.QInputDialog, 'getText')
+
+    editor.go_to_line(1)
+    block = editor.textCursor().block()
+
+    # Test with default call to slot 1 on text line containing code.
+    reset_emits(editor)
+    ab(1)
+    editor_assert_helper(editor, block, bm=[(1, 0)], emits=True)
+
+    # Test on indented line and add multiple bookmarks.
+    reset_emits(editor)
+    editor.go_to_line(4)
+    block = editor.textCursor().block()
+    ab(1)
+    cursor = editor.textCursor()
+    cursor.movePosition(QTextCursor.Right, n=2)
+    editor.setTextCursor(cursor)
+    ab(2)
+    editor_assert_helper(editor, block, bm=[(1, 0), (2, 2)], emits=True)
+
+
+def test_get_bookmarks(code_editor_bot):
+    """Test CodeEditor.get_bookmarks. Returns data found in textblocks."""
+    editor, _ = code_editor_bot
+    gb = editor.get_bookmarks
+
+    assert(gb() == {})
+
+    # Add bookmarks.
+    bm = {1: ('filename', 1, 0), 2: ('filename', 3, 5), 3: ('filename', 4, 3)}
+    editor.set_bookmarks(bm)
+    assert(gb() == {1: [1, 0], 2: [3, 5], 3: [4, 3]})
+
+
+def test_clear_bookmarks(code_editor_bot):
+    """Test CodeEditor.clear_bookmarks. Removes bookmarks from all blocks."""
+    editor, _ = code_editor_bot
+
+    assert len(editor.blockuserdata_list) == 0
+
+    bm = {1: ('filename', 1, 0), 2: ('filename', 3, 5)}
+    editor.set_bookmarks(bm)
+    assert editor.get_bookmarks() == {1: [1, 0], 2: [3, 5]}
+    assert len(editor.blockuserdata_list) == 2
+
+    editor.clear_bookmarks()
+    assert editor.get_bookmarks() == {}
+    # Even though there is a 'del data' that would pop the item from the
+    # list, the __del__ funcion isn't called.
+    assert len(editor.blockuserdata_list) == 2
+    for data in editor.blockuserdata_list:
+        assert not data.bookmarks
+
+
+def test_update_bookmarks(code_editor_bot):
+    """Test CodeEditor.update_bookmarks. Check if signal is emitted."""
+    editor, _ = code_editor_bot
+    reset_emits(editor)
+    editor.bookmarks_changed.emit.assert_not_called()
+    # update_bookmarks is the slot for the blockCountChanged signal.
+    editor.textCursor().insertBlock()
+    editor.bookmarks_changed.emit.assert_called_with()
+
+
+def test_save_bookmark(setup_editor):
+    """Test Plugin.save_bookmark. 
+
```

Comment:
Our agreed convention is to make the summary line on its own line for multi-line docstrings